### ISSUE TITLE
How to backpropagate from rendered image to input mesh vertices

### ISSUE THREAD

**Issue by karta2155802** [facebookresearch/pytorch3d#1432]
Hi, I'm trying to render a set of vertices estimated by other neural network along with their mesh faces to a mask and compare to the groundtruth mask by the code below.
I switch all input data to random number instead for simplicity.
```python
import torch
import torch.nn as nn
from torch.nn import functional as F
from pytorch3d.structures import Meshes
# rendering components
from pytorch3d.renderer import (
    RasterizationSettings, MeshRenderer, MeshRasterizer, 
    HardPhongShader, PerspectiveCameras, TexturesVertex
)

def render_mask(verts, faces, focal, princpt, image_size):
    B = verts.size(0)
    device = verts.device
    verts_rgb = torch.ones_like(verts)
    textures=TexturesVertex(verts_features=verts_rgb)
    mesh = Meshes(
        verts=[v for v in verts],
        faces=[f for f in faces],
        textures=textures
    )

    cam_img_size = [image_size for i in range(B)]
    cameras = PerspectiveCameras(device=device, focal_length=focal, principal_point=princpt, in_ndc=False, 
        image_size=cam_img_size)
    raster_settings = RasterizationSettings(image_size=image_size, blur_radius=0.0, faces_per_pixel=1)
    renderer = MeshRenderer(
        rasterizer=MeshRasterizer(cameras=cameras, raster_settings=raster_settings), 
        shader=HardPhongShader(device=device, cameras=cameras)
    )

    images = renderer(mesh)
    
    masks = (images[...,3]).unsqueeze(1) # take the alpha channel
    return masks


# settings
B = 4 # batch size
focal = torch.randn(B,2)
princpt = torch.randn(B,2)
img_size = (10,10)
faces = torch.tensor([[1,2,3],[2,3,5],[4,7,8],[5,7,9]])
faces = faces.repeat(B,1,1)
gt = torch.randn(B,1,10,10) # groundtruth

# input vertices which are estimated from other network, here I use random number instead
verts = torch.randn((B,10,3), requires_grad=True)
# get rendered mask
mask = render_mask(verts, faces, focal, princpt, img_size)

mask.retain_grad()
verts.retain_grad()

loss = F.binary_cross_entropy(mask, gt)
print('loss:', loss)
loss.backward()
print('mask:', mask.grad.sum())
print('verts:', verts.grad.sum())
'''
outputs
  loss: tensor(-4.9824, grad_fn=<BinaryCrossEntropyBackward0>)
  mask: tensor(1.4741e+11)
  verts: tensor(0.)
'''

```
I expect there will be gradients for the input vertices, but they end up being zeros
Is it possible to optimize the mesh vertices through such procedure?
And why the grad for mask is extremely large?

**Comment by bottler** [comment]
You might need a different shader, like SoftPhongShader, to have gradients with respect to positions. Large loss may be because binary_cross_entropy will behave oddly when values are outside `[0,1]`. Can you try getting `gt` from `torch.rand` instead of `randn`?

**Comment by karta2155802** [comment]
Thanks for the reply.
After I tried your suggestion, the problems still remains, `0` gradients for `verts`.
And I had confirm that inputs for BCE are in range `[0, 1]`, but the sum of `.grad` is still large.
